Add release procedure and release script (#1127)
eth-cscs/DLA-Future: RELEASE_PROCEDURE.md

# Release procedure for DLA-Future

DLA-Future follows [Semantic Versioning](https://semver.org).

1. For minor and major releases: create and check out a new branch at an
   appropriate point on `main` with the name `release-major.minor`.  `major` and `minor` should be the
   major and minor versions of the release. For patch releases: check out the corresponding
   `release-major.minor` branch.

> 1. Write release notes in `CHANGELOG.md`. Check for issues and pull requests for the release on the

Reviewer: ```suggestion
2. Write release notes in `CHANGELOG.md`. Check for issues and pull requests for the release on the
```
? Same for the other points.

Author: I know it looks funny, but this is actually rendered with correctly numbered items on GitHub (https://github.com/eth-cscs/DLA-Future/blob/369055535bd658c2151ccbba1a63e27e45c55f0b/RELEASE_PROCEDURE.md). I prefer this to explicitly having to renumber items if some are deleted/added. But then I'm not sure where this is most likely to be viewed (in an editor or rendered).

Reviewer: That's neat!

>    [DLA-F Planning board](https://github.com/orgs/eth-cscs/projects/1). All list entries and paragraphs

Reviewer: ```suggestion
   [DLA-Future Planning board](https://github.com/eth-cscs/DLA-Future/projects/2). All list entries and paragraphs
```

Author: The on you link to is the "old" project board. Perhaps we can close or delete it now? @rasolca do you mind if we close it?

`DLA-F Planning board` matches the title of the new project board, but maybe we just update the title there to be `DLA-Future Planning board`. Seems more consistent?

Reviewer: I see. The other is a dangling link though. Maybe I don't have the right permissions? Happy to leave `DLA-F` if that's the actual name of the board (which I was not able to access).

Author: Aha, yes, I just checked and it was private, not including you! Sorry about that... Can you see it now? I made it public, since it just contains all the issues/PRs that we anyway have publicly visible.

>    must be on a single line for correct formatting on GitHub.

Reviewer 2: I would add a note like "Only relevant changes should be mentioned in the changelog."

Author: Do you have a definition for "relevant changes"? Only user-facing changes? Only major changes? Something else?

Reviewer 2: A sort of guideline (but for sure non exhaustive) can be

I excluded:
 - anything that was related to CI and github actions
 - anything related to spack package (Note: change of dependencies are not stated as spack package change)
 - documentation related changes
 - anything related to plotting scripts
Must be included:
 - changes in user API
 - new functionalities
 - bugfixes
 - performance critical changes
 - change in dependencies

Author: How does this look: https://github.com/eth-cscs/DLA-Future/pull/1127/commits/070aa621075f9d0d305294304f0297e407b7ef66?

1. Make sure the version is set correctly in `CMakeLists.txt`.

1. When making a post-1.0.0 major release, remove deprecated functionality if
   appropriate.

1. Update the minimum required versions if necessary.

> 1. Add a link to the documentation for the release. The documentation will be generated automatically after the release has been made.

Author: ```suggestion
1. Add a link to the documentation for the release. The documentation will be generated automatically after the `vx.y.z` tag is created and pushed.
```
?

> 1. Add a link to the documentation for the release. The documentation will be generated automatically after the release has been made.

Author: https://github.com/eth-cscs/DLA-Future/pull/1127/commits/9a0b05c6d1a367789431efe4b33bc72f69db9475?

1. Create a release on GitHub using the script `scripts/roll_release.sh`. This
   script automatically tags the release with the corresponding release number.  You'll need to set
   `GITHUB_TOKEN` or both `GITHUB_USER` and `GITHUB_PASSWORD` for the hub release command. When creating
   a `GITHUB_TOKEN`, the only access necessary is `public_repo`.

> 1. Merge release branch into `master` (with `--no-ff`).

Reviewer 2: This might not work if some fixes has to be adapted to apply to the previous release.

Author: Yep, fair enough. Update according to https://github.com/eth-cscs/DLA-Future/pull/1128#issuecomment-2092480407?

Author: Removed this step in https://github.com/eth-cscs/DLA-Future/pull/1127/commits/21df7ac30d396152c27db72e34c462cf8c0279c0.

1. Modify the release procedure if necessary.

> 1. Synchronize [upstream spack
>    package](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dla-future/package.py)
>    with local repository.

Reviewer 2: ```suggestion
1. Update spack recipe in `/spack/packages/dla-future/package.py` adding the new release.

1. Synchronize [upstream spack
   package](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dla-future/package.py)
   with local repository.
```

>    with local repository.

Reviewer 2: Maybe add a note about checking the comments for the functionalities that should not go in spack?

Author: https://github.com/eth-cscs/DLA-Future/pull/1127/commits/0feb9309b5f869dd8cd502cf19bd7fe76e341748?
